Code/DistGeom/TriangleSmooth.cpp

```diff
@@ -16,61 +15,52 @@ bool triangleSmoothBounds(BoundsMatPtr boundsMat, double tol) {
   return triangleSmoothBounds(boundsMat.get(), tol);
 }
 bool triangleSmoothBounds(BoundsMatrix *boundsMat, double tol) {
-  int npt = boundsMat->numRows();
-  int i, j, k;
-  double Uik, Lik, Ukj, sumUikUkj, diffLikUjk, diffLjkUik;
-
-  for (k = 0; k < npt; k++) {
-    for (i = 0; i < npt - 1; i++) {
+  auto npt = boundsMat->numRows();
+  for (auto k = 0u; k < npt; k++) {
+    for (auto i = 0u; i < npt - 1; i++) {
       if (i == k) {
         continue;
       }
-      Uik = boundsMat->getUpperBound(i, k);
-      Lik = boundsMat->getLowerBound(i, k);
-      for (j = i + 1; j < npt; j++) {
+      auto ii = i;
+      auto ik = k;
+      if (ii > ik) {
+        std::swap(ii, ik);
+      }
+
+      const auto Uik = boundsMat->getValUnchecked(ii, ik);  // upper bound
+      const auto Lik = boundsMat->getValUnchecked(ik, ii);  // lower bound
+      for (auto j = i + 1; j < npt; j++) {
         if (j == k) {
           continue;
         }
-        Ukj = boundsMat->getUpperBound(k, j);
-        sumUikUkj = Uik + Ukj;
-        if (boundsMat->getUpperBound(i, j) > sumUikUkj) {
-          boundsMat->setUpperBound(i, j, sumUikUkj);
+        auto jj = j;
+        auto jk = k;
+        if (jj > jk) {
+          std::swap(jj, jk);
         }
-
-        diffLikUjk = Lik - Ukj;
-        diffLjkUik = boundsMat->getLowerBound(j, k) - Uik;
-        if (boundsMat->getLowerBound(i, j) < diffLikUjk) {
-          boundsMat->setLowerBound(i, j, diffLikUjk);
-        } else if (boundsMat->getLowerBound(i, j) < diffLjkUik) {
-          boundsMat->setLowerBound(i, j, diffLjkUik);
+        const auto Ukj = boundsMat->getValUnchecked(jj, jk);  // upper bound
+        const auto sumUikUkj = Uik + Ukj;
+        if (boundsMat->getValUnchecked(i, j) > sumUikUkj) {
+          // adjust the upper bound
+          boundsMat->setValUnchecked(i, j, sumUikUkj);
         }
-        double lBound = boundsMat->getLowerBound(i, j);
-        double uBound = boundsMat->getUpperBound(i, j);
-        // std::cerr << "     smooth: " << i << "-" << j << "(" << k
-        //           << "): " << boundsMat->getLowerBound(i, j) << " "
-        //           << boundsMat->getUpperBound(i, j) << " lik: " << Lik
-        //           << " uik: " << Uik
-        //           << " lkj: " << boundsMat->getLowerBound(j, k)
-        //           << " ukj: " << Ukj << "\n";
 
+        const auto diffLikUjk = Lik - Ukj;
+        const auto diffLjkUik = boundsMat->getValUnchecked(jk, jj) - Uik;
+        if (boundsMat->getValUnchecked(j, i) < diffLikUjk) {
+          // adjust the lower bound
+          boundsMat->setValUnchecked(j, i, diffLikUjk);
```

Comment:
Are these guaranteed to be non negative now?  diffLikUjk and diffLikUik

Reply:
I think they are, but might be worth a comment.

Reply:
I don't understand the question. Why is the sign of those values significant?

Reply:
The setUpper/LowerBounds had this invariant check which no longer exists.

```
CHECK_INVARIANT(val >= 0.0, "Negative upper bound");
```

Reply:
> The setUpper/LowerBounds had this invariant check which no longer is called.
> 
> 
> 
> ```
> 
> CHECK_INVARIANT(val >= 0.0, "Negative upper bound");
> 
> ```

I don't think this is a concern since we are in the triangle bounds smoothing code. It's not random code updating the bounds matrices.

Reply:
I agree with that assessment.